File: java/src/main/java/Day03_Mull_It_Over/Day3_Part2.java
package Day03_Mull_It_Over;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


public class Day3_Part2 {
    private static final Pattern MUL_PATTERN = Pattern.compile("mul\\((\\d+),(\\d+)\\)|do\\(\\)|don't\\(\\)");

    public static void main(String[] args) {
        boolean enabled = true;
        int totalSum = 0;

        String filePath = Objects.requireNonNull(
                Day3_Part2.class.getResource("/Day3_input.txt")
        ).getPath();
        filePath = URLDecoder.decode(filePath, StandardCharsets.UTF_8);

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;

            while ((line = br.readLine()) != null) {
                Matcher matcher = MUL_PATTERN.matcher(line);

                while (matcher.find()) {
                    if (matcher.group(0).equals("do()")) enabled = true;
                    else if (matcher.group(0).equals("don't()")) enabled = false;
                    else if (enabled) {
                        int num1 = Integer.parseInt(matcher.group(1));
                        int num2 = Integer.parseInt(matcher.group(2));

                        totalSum += num1 * num2;
                    }
                }
            }
        } catch (IOException e) {
            System.out.println("File not found");
        }

        System.out.println("Sum of products of `mul` instructions: " + String.format("%,d", totalSum));
    }
}

File: python/03-Mull_It_Over/Day3_Part2.py
# Advent of Code 2024
# Solution - Day 3, Part 2 - Mull It Over
# Author: CSK

import re

MUL_PATTERN = re.compile(r"mul\((\d+),(\d+)\)|do\(\)|don't\(\)")

def main():
    total_sum: int = 0
    enabled: bool = True

    with open("input.txt", "r", encoding="utf-8") as file:
        for line in file:
            for match in MUL_PATTERN.finditer(line):
                if match.group(0) == "do()":
                    enabled = True
                elif match.group(0) == "don't()":
                    enabled = False
                elif enabled:
                    total_sum += int(match.group(1)) * int(match.group(2))

    print(f"Sum of products of `mul` instructions: {total_sum:,}")

if __name__ == "__main__":
    main()
